lib/activerecord_sqlserver_adapter/jdbc/dataset.rb

```diff
@@ -0,0 +1,225 @@
+          c
+        end
+        # :nocov:
+      end
+
+      # Set the db, opts, and cache for the copy of the dataset.
+      def initialize_copy(c)
+        @db = c.db
+        @opts = Hash[c.opts]
+        if cols = c.cache_get(:_columns)
+          @cache = {:_columns=>cols}
+        else
+          @cache = {}
+        end
+      end
+
+      def with_sql(sql, opts=OPTS)
+        clone(opts.merge(:sql => sql))
+      end
+
+      def each(options=OPTS)
+        fetch_rows(opts[:sql], opts.merge(options)){|r| yield r}
+        self
+      end
+
+      alias :all :entries
+
+      # Correctly return rows from the database and return them as hashes.
+      def fetch_rows(sql, opts=OPTS, &block)
+        db.execute(sql){|result| process_result_set(result, opts, &block)}
+        self
+      end
+
+      # Split out from fetch rows to allow processing of JDBC result sets
+      # that don't come from issuing an SQL string.
+      def process_result_set(result, opts=OPTS)
+        meta = result.getMetaData
+        if fetch_size = opts[:fetch_size]
+          result.setFetchSize(fetch_size)
+        end
+
+        converters = []
+        self.columns = meta.getColumnCount.times.map do |i|
+          col = i + 1
+          converters << TypeConverter::MAP[meta.getColumnType(col)]
+          meta.getColumnLabel(col)
+        end
+
+        fetch_as_array = opts[:as] == :array
+        while result.next
+          row = fetch_as_array ? [] : {}
+          _columns.each_with_index do |column, i|
+            k = fetch_as_array ? i : column
+            col = i+1
+            row[k] = converters[i].call(result, col, opts)
+          end
+          yield row
+        end
+      ensure
+        result.close
+      end
+
+      # Returns the columns in the result set in order as an array of symbols.
+      # If the columns are currently cached, returns the cached value. Otherwise,
+      # a SELECT query is performed to retrieve a single row in order to get the columns.
+      #
+      #   DB.fetch('SELECT 1 AS a, 2 AS b').columns
+      #   # => [:a, :b]
+      def columns
+        _columns || columns!
+      end
+
+      # Ignore any cached column information and perform a query to retrieve
+      # a row in order to get the columns.
+      #
+      #   dataset = DB.fetch('SELECT 1 AS a, 2 AS b')
+      #   dataset.columns!
+      #   # => [:a, :b]
+      def columns!
+        ds = clone(opts.merge(:sql => "SELECT TOP 1 [T1].* FROM (#{opts[:sql]}) \"T1\""))
+        ds.each{break}
+
+        if cols = ds.cache[:_columns]
+          self.columns = cols
+        else
+          []
+        end
+      end
+
+      protected
+
+      # Access the cache for the current dataset.  Should be used with caution,
+      # as access to the cache is not thread safe without a mutex if other
+      # threads can reference the dataset.  Symbol keys prefixed with an
+      # underscore are reserved for internal use.
+      attr_reader :cache
```

Comment:
The cache is hit frequently. It saves one query each time which is needed to fetch the column information. The column information is used for the type conversion from java to ruby types and for the column names of the results dataset.

Maybe we can streamline the code none the less. But I would first like to get the test suite completely green.